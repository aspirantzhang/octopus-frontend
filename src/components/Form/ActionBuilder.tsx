import React from 'react';
import { Button } from 'antd';
import { ButtonType } from 'antd/lib/button';

export const ActionBuilder = (
  actions: PageAPI.ActionData[],
  actionHandler: API.ActionHandler,
  loading: boolean = false,
) => {
  return actions.map((action: PageAPI.ActionData) => {
    switch (action.component) {
      case 'button':
        return (
          <Button
            type={action.type as ButtonType}
            key={action.action}
            loading={loading}
            onClick={() => {
              actionHandler(action);
            }}
            id={action.id || ''}
          >
            {action.text}
          </Button>
        );

      default:
        return null;
    }
  });
};
