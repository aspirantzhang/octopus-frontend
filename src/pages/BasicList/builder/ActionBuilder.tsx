import React from 'react';
import { Button } from 'antd';
import type { ButtonType } from 'antd/lib/button';
import { ThunderboltTwoTone } from '@ant-design/icons';

const ActionBuilder = (
  actions: BasicListApi.Action[] | undefined,
  actionHandler: BasicListApi.ActionHandler,
  loading = false,
  record?: BasicListApi.Field,
) => {
  return (Array.isArray(actions) ? actions : []).map((action) => {
    if (action.name === 'quick_edit') {
      return (
        <Button
          key={action.name}
          type="dashed"
          icon={<ThunderboltTwoTone />}
          onClick={() => {
            actionHandler(action, record);
          }}
          loading={loading}
          className={`${action.name.replace(/[^\w\d]/g, '-')}-btn`}
        ></Button>
      );
    }

    return (
      <Button
        key={action.name}
        type={action.type as ButtonType}
        onClick={() => {
          actionHandler(action, record);
        }}
        loading={loading}
        className={`${action.name.replace(/[^\w\d]/g, '-')}-btn`}
      >
        {action.title}
      </Button>
    );
  });
};
export default ActionBuilder;
