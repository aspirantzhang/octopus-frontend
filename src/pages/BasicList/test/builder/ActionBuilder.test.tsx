/**
 * @jest-environment jsdom
 */
import Renderer from 'react-test-renderer';
import '@testing-library/jest-dom/extend-expect';
import { render, screen, fireEvent } from '@testing-library/react';
import ActionBuilder from '../../builder/ActionBuilder';

const validActions = [
  {
    component: 'button',
    type: 'primary',
    name: 'edit',
    title: 'Edit',
    call: 'modal',
    method: null,
    uri: '/api/admins/:id',
  },
  {
    component: 'button',
    type: 'default',
    name: 'delete',
    title: 'Delete',
    call: 'delete',
    method: 'post',
    uri: '/api/admins/delete',
  },
];
const invalidActions = [
  {
    component: 'invalidComponent',
    type: 'primary',
    name: 'edit',
    title: 'Edit',
    call: 'modal',
    method: null,
    uri: '/api/admins/:id',
  },
];

describe('ActionBuilder', () => {
  test('snapshot test', () => {
    const dom = Renderer.create(ActionBuilder(validActions as any, () => {}) as any).toJSON();
    expect(dom).toMatchSnapshot();
  });

  test('valid params', () => {
    render(ActionBuilder(validActions as any, () => {}) as any);
    expect(screen.getByText('Delete')).toBeInTheDocument();
    expect(screen.getByText('Edit')).toBeInTheDocument();
    fireEvent.click(screen.getByText('Edit'));
  });

  test('invalid params', () => {
    expect(ActionBuilder(null as any, () => {})).toEqual([]);
    expect(ActionBuilder([] as any, () => {})).toEqual([]);
    expect(ActionBuilder({} as any, () => {})).toEqual([]);
    expect(ActionBuilder(undefined as any, () => {})).toEqual([]);
    expect(ActionBuilder(NaN as any, () => {})).toEqual([]);
    expect(ActionBuilder(true as any, () => {})).toEqual([]);
    expect(ActionBuilder(false as any, () => {})).toEqual([]);
    expect(ActionBuilder('' as any, () => {})).toEqual([]);
    expect(ActionBuilder(invalidActions as any, () => {})).toEqual([null]);
  });
});
