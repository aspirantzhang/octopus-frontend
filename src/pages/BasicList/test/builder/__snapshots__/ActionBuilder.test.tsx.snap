// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActionBuilder snapshot test 1`] = `
Array [
  <button
    className="ant-btn ant-btn-primary edit-btn"
    onClick={[Function]}
    type="button"
  >
    <span>
      Edit
    </span>
  </button>,
  <button
    className="ant-btn ant-btn-default delete-btn"
    onClick={[Function]}
    type="button"
  >
    <span>
      Delete
    </span>
  </button>,
]
`;
